relay-server/src/middlewares/body_timing.rs

```diff
@@ -0,0 +1,288 @@
+/// - failed during reading ("failed")
+/// - was dropped before completion ("dropped")
+///
+/// The reported size reflects the actual number of bytes read, which may differ from
+/// the `Content-Length` header. For bodies that fail or are dropped mid-reading,
+/// the size represents only the bytes successfully read before the interruption.
+#[derive(Clone, Debug)]
+pub struct BodyTimingLayer;
+
+impl BodyTimingLayer {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl<S> Layer<S> for BodyTimingLayer {
+    type Service = BodyTiming<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        BodyTiming::new(inner)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct BodyTiming<S> {
+    inner: S,
+}
+
+impl<S> BodyTiming<S> {
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S> Service<Request<Body>> for BodyTiming<S>
+where
+    S: Service<Request<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let (parts, body) = req.into_parts();
+        let timed_body = Body::new(TimedBody::new(body, parts.uri.to_string()));
+        let request = Request::from_parts(parts, timed_body);
+        self.inner.call(request)
+    }
+}
+
+#[derive(Debug)]
+struct TimedBody {
+    inner: Body,
+    reading_started_at: Option<Instant>,
+    size: u32,
+    route: String,
+}
+
+impl TimedBody {
+    fn new<T>(inner: Body, route: T) -> Self
+    where
+        T: Into<String>,
+    {
+        Self {
+            inner,
+            reading_started_at: None,
+            size: 0,
+            route: route.into(),
+        }
+    }
+
+    fn finish_timing(&mut self, status: &str) {
+        if let Some(started_at) = self.reading_started_at.take() {
+            self.emit_metric(started_at.elapsed(), status);
+        }
+    }
+
+    fn emit_metric(&self, duration: Duration, status: &str) {
+        metric!(
+            timer(RelayTimers::BodyReading) = duration,
+            route = &self.route,
+            size = size_category(self.size),
+            status = status
+        )
+    }
+}
+
+impl hyper::body::Body for TimedBody {
+    type Data = <Body as HttpBody>::Data;
+    type Error = <Body as HttpBody>::Error;
+
+    fn poll_frame(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+        if self.reading_started_at.is_none() {
+            self.reading_started_at = Some(Instant::now());
+        }
+        let pinned = Pin::new(&mut self.inner);
+        let poll_result = pinned.poll_frame(cx);
+
+        if let Poll::Ready(Some(Ok(frame))) = &poll_result {
+            if let Some(data) = frame.data_ref() {
+                self.size += data.len() as u32;
```

Comment:
Casting a `usize` to `u32` may wrap, you can just count the size as `usize` and not cast.